File: golang/1249.minimum-remove-to-make-valid-parentheses.go
func minRemoveToMakeValid(s string) string {
	tmp := []byte(s)
	stack := make([]int, 0, len(s)/2)

	for i := 0; i < len(s); i++ {
		if s[i] == '(' {
			stack = append(stack, i)
		} else if s[i] == ')' {
			if len(stack) > 0 && s[stack[len(stack)-1]] == '(' {
				stack = stack[:len(stack)-1]
			} else {
				stack = append(stack, i)
			}
		}
	}

	for len(stack) > 0 {
		tmp[stack[len(stack)-1]] = 0
		stack = stack[:len(stack)-1]
	}

	res := make([]byte, 0, len(s))
	for i := 0; i < len(tmp); i++ {
		if tmp[i] != 0 {
			res = append(res, tmp[i])
		}
	}

	return string(res)
}

File: cpp/1249.minimum-remove-to-make-valid-parentheses.cpp
#include <stack>
#include <string>
using namespace std;

class Solution {
public:
  string minRemoveToMakeValid(string s) {
    stack<int> st;
    string tmp(s.begin(), s.end());

    for (int i = 0; i < s.size(); i++) {
      if (s[i] == '(')
        st.push(i);
      else if (s[i] == ')') {
        if (!st.empty() && s[st.top()] == '(')
          st.pop();
        else
          st.push(i);
      }
    }

    while (!st.empty()) {
      tmp[st.top()] = '*';
      st.pop();
    }

    string result;
    result.reserve(tmp.size());

    for (char c : tmp) {
      if (c == '*')
        continue;
      result.push_back(c);
    }

    return result;
  }
};